README.md
![NTELLI](/images/banner.png)

<h1 style="font-size:2rem; width:100%; text-align:center; margin-bottom: 1rem"}> NTELLI Automatic Unit Test</h1>

Automatically generate Jest unit test files and get code refactoring suggestions with our extension. Streamlined with chatGPT turbo 3.5, this extension makes the process of writing comprehensive test cases a breeze. Simply activate the extension on your active file and let it generate a test file for you. Highlight a section of code and get suggestions on how to refactor it. Never worry about writing lengthy test cases again.

![Alt Text](/images/create.gif)

## Commands

_Open command palette_

CMD + shift + p

Write **ncut** = NTELLI CREATE UNIT TEST

Write **nrut** = NTELLU REFACTOR UNIT TEST

## Settings

Open up the extension's settings and fill in the API key and organization ID, that you will get from your openai account.


CHANGELOG.md
# Change Log

All notable changes to the "ntelli-aut" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/stream.ts
import { TransformCallback } from 'stream';
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import axios from 'axios';

const stream = async (folderPath: string, fileName: string, prompt: string) => {
	const streamNode = require('stream');
	const vsconfig = vscode.workspace.getConfiguration();
	let OPENAI_APIKEY = vsconfig.get('openai.api_key') as string;
	let OPENAI_ORGANIZATIONID = vsconfig.get('openai.organization_ID') as string;
	if (!OPENAI_APIKEY || !OPENAI_ORGANIZATIONID) {
		vscode.window.showErrorMessage('Please check your settings, apiKey and organizationID are required.');
		return;
	}

	let testFileName = '';
	if (fileName.includes('ts')) testFileName = `${fileName.split('.').slice(0, -1).join('.')}.test.ts`;
	if (fileName.includes('tsx')) testFileName = `${fileName.split('.').slice(0, -1).join('.')}.test.tsx`;

	const testFilePath = path.join(folderPath, testFileName);
	fs.writeFileSync(testFilePath, '');

	function commentTextBlocks() {
		fs.readFile(testFilePath, 'utf8', (err, data) => {
			if (err) {
				console.error(err);
				return;
			}
			const commentedBlock = data.replace(
				// /(^[A-Z][\w\s,.;!?-]*[.:])|([^\n]*[.:][\s\n]*(?=[A-Z]|[0-9]|\(|\s*$))/gm,
				/^[A-Z][^.\n]*[:.]\s*/gm,

				match => {
					if (data.indexOf(match) >= data.indexOf('`') && data.indexOf(match) <= data.lastIndexOf('`')) {
						return match;
					} else {
						return `// ${match}`;
					}
				}
			);

			fs.writeFileSync(testFilePath, commentedBlock);
		});
	}

	const transformStream = new streamNode.Transform({
		transform(chunk: Buffer | string, encoding: string | null, callback: TransformCallback) {
			let transformedData = chunk.toString('utf-8');
			transformedData = transformedData.replace(/\\n/g, '\n');
			transformedData = transformedData.replace(/`/g, '');

			const regex = /"content":"([^"]*)"/;
			const match = regex.exec(transformedData);

			let content = match ? match[1] : '';

			this.push(content);
			callback();
		},
	});

	try {
		const url = 'https://api.openai.com/v1/chat/completions';
		const headers = {
			'Content-Type': 'application/json',
			Authorization: `Bearer ${OPENAI_APIKEY ?? ''}`,
		};
		const payload = {
			model: 'gpt-3.5-turbo',
			messages: [
				{
					role: 'system',
					content:
						'You are an expert in Jest unit tests. Generate a Jest unit test based on the input with 100% test coverage. Do not use Enzyme, Mocha, Jasmine or any other javascript test framework in the tests.',
				},
				{ role: 'user', content: prompt },
			],
			stream: true,
		};
		axios
			.post(url, payload, {
				headers: headers,
				responseType: 'stream',
			})
			.then(response => {
				response.data.pipe(transformStream).on('data', (chunk: any) => {
					fs.appendFileSync(testFilePath, chunk);
				});

				response.data.on('end', () => {
					commentTextBlocks();
					vscode.window.showInformationMessage(`Unit test generated and saved at ${testFilePath}`);
				});
			})
			.catch(error => {
				console.error(error);
			});
	} catch (error: any) {
		if (error.response?.status === 401) {
			vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
		} else {
			vscode.window.showErrorMessage(error.message);
		}
	}
};

export const createTest = vscode.commands.registerCommand('ntelli-aut.createTest', async () => {
	try {
		const editor = vscode.window.activeTextEditor;
		if (editor) {
			const folderPath = path.dirname(editor.document.fileName);
			const fileName = path.basename(editor.document.fileName);
			const prompt = editor.document.getText();
			const cancellationTokenSource = new vscode.CancellationTokenSource();
			await vscode.window.withProgress(
				{
					location: vscode.ProgressLocation.Notification,
					title: 'Generating unit test...',
					cancellable: true,
				},
				async (progress, token) => {
					token.onCancellationRequested(() => {
						console.log('User canceled the long running operation');
						cancellationTokenSource?.cancel();
					});
					await stream(folderPath, fileName, prompt);
					progress.report({ increment: 100 });
				}
			);
		}
	} catch (error: any) {
		if (error.response?.status === 401) {
			vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
		} else {
			vscode.window.showErrorMessage(error.message);
		}
	}
});


src/refactor.ts
import { TransformCallback } from 'stream';
import * as vscode from 'vscode';
import axios from 'axios';

async function getRefactorSuggestion(): Promise<void> {
	const streamNode = require('stream');
	const vsconfig = vscode.workspace.getConfiguration();
	let OPENAI_APIKEY = vsconfig.get('openai.api_key') as string;
	let OPENAI_ORGANIZATIONID = vsconfig.get('openai.organization_ID') as string;
	if (!OPENAI_APIKEY || !OPENAI_ORGANIZATIONID) {
		vscode.window.showErrorMessage('Please check your settings, apiKey and organizationID are required.');
		return;
	}

	// Get the current text editor
	let editor = vscode.window.activeTextEditor;
	if (editor) {
		// Get the position of the last line of the highlighted text
		let selection = editor.selection;

		// Modify the prompt to include the cursor position
		const input = editor.document.getText(
			new vscode.Range(selection.start.line + 0, 0, selection.end.line, selection.end.character)
		);

		const transformStream = new streamNode.Transform({
			transform(chunk: Buffer | string, encoding: string | null, callback: TransformCallback) {
				let transformedData = chunk.toString('utf-8');
				transformedData = transformedData.replace(/\\n/g, '\n');
				transformedData = transformedData.replace(/`/g, '');
				transformedData = transformedData.replace(/\\t/g, '');

				const regex = /"content":"([^"]*)"/;
				const match = regex.exec(transformedData);

				let content = match ? match[1] : '';

				this.push(content);
				callback();
			},
		});

		try {
			const url = 'https://api.openai.com/v1/chat/completions';
			const headers = {
				'Content-Type': 'application/json',
				Authorization: `Bearer ${OPENAI_APIKEY ?? ''}`,
			};
			const payload = {
				model: 'gpt-3.5-turbo',
				messages: [
					{
						role: 'system',
						content:
							'You are an expert in refactoring Jest unit tests. Come with your best opnion on how to refactor the provided input, with comments and code.',
					},
					{ role: 'user', content: input },
				],
				stream: true,
			};
			axios
				.post(url, payload, {
					headers: headers,
					responseType: 'stream',
				})
				.then(response => {
					const view = vscode.window.createWebviewPanel(
						'ntelli-aut.refactorTest.results',
						'Refactor Test Results',
						vscode.ViewColumn.One,
						{
							enableScripts: true,
						}
					);

					let text = '';
					response.data.pipe(transformStream).on('data', (chunk: any) => {
						text += chunk;
						if (view) {
							view.webview.html = `
							<!DOCTYPE html>
							<html>
								<head>
									<style>
										body {
											color: #f2f2f2;
											font-size: 16px;
											line-height: 1.5;
											padding: 20px;
										}
										p {
											margin-bottom: 20px;
										}
									</style>
								</head>
							
								<body>
									<p>${text}</p>
								</body>
							</html>
							`;
						}
					});

					response.data.on('end', () => {
						vscode.window.showInformationMessage(`Refactor suggestion has been provided`);
					});
				})
				.catch(error => {
					console.error(error);
				});
		} catch (error: any) {
			if (error.response.status === 401) {
				vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
			} else {
				vscode.window.showErrorMessage(error.message);
			}
		}
	} else {
		vscode.window.showErrorMessage('No text editor is active. Please open a file first.');
	}
}

export const refactorTest = vscode.commands.registerCommand('ntelli-aut.refactorTest', async () => {
	try {
		await vscode.window.withProgress(
			{
				location: vscode.ProgressLocation.Notification,
				title: 'Loading...',
				cancellable: false,
			},
			async (progress, token) => {
				token.onCancellationRequested(() => {
					console.log('User canceled the long running operation');
				});
				await getRefactorSuggestion();
				progress.report({ increment: 100 });
			}
		);
	} catch (err: any) {
		if (err.code === 'ENOENT') {
			vscode.window.showErrorMessage('settings.json file not found.');
		} else if (err instanceof SyntaxError) {
			vscode.window.showErrorMessage('settings.json file is not a valid JSON file.');
		} else {
			vscode.window.showErrorMessage('An error occurred while reading settings.json file.');
		}
	}
});


src/generate.ts
import * as vscode from 'vscode';
import { Configuration, OpenAIApi } from 'openai';
import * as path from 'path';
import * as fs from 'fs';

async function generateTestFile(folderPath: string, fileName: string, prompt: string) {
	const vsconfig = vscode.workspace.getConfiguration();
	let OPENAI_APIKEY = vsconfig.get('openai.api_key') as string;
	let OPENAI_ORGANIZATIONID = vsconfig.get('openai.organization_ID') as string;
	if (!OPENAI_APIKEY || !OPENAI_ORGANIZATIONID) {
		vscode.window.showErrorMessage('Please check your settings, apiKey and organizationID are required.');
		return;
	}

	const configuration = new Configuration({
		organization: OPENAI_ORGANIZATIONID,
		apiKey: OPENAI_APIKEY,
	});

	let openai = new OpenAIApi(configuration);
	try {
		const response = await openai.createChatCompletion({
			model: 'gpt-3.5-turbo',
			messages: [
				{
					role: 'system',
					content:
						'You are an expert in jest unit tests. Generate a jest unit test based on the input with 100% test coverage',
				},
				{ role: 'user', content: prompt },
			],
			temperature: 0,
			n: 1,
		});
		let generatedTest = response.data?.choices?.[0]?.message?.content;
		if (typeof generatedTest === 'string') {
			// Remove ``` from the code block
			generatedTest = generatedTest.replace(/`/g, '');

			const testFileName = `${fileName.split('.').slice(0, -1).join('.')}.test.ts`;
			const testFilePath = path.join(folderPath, testFileName);
			fs.writeFileSync(testFilePath, generatedTest.toString());

			vscode.window.showInformationMessage(`Unit test generated and saved at ${testFilePath}`);
		} else {
			vscode.window.showErrorMessage('Failed to generate the unit test. Please check your OpenAI response.');
		}
	} catch (error: any) {
		if (error.response?.status === 401) {
			vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
		} else {
			vscode.window.showErrorMessage(error.message);
		}
	}
}

export const generateUnitTest = vscode.commands.registerCommand('ntelli-aut.generateUnitTest', async () => {
	try {
		const editor = vscode.window.activeTextEditor;
		if (editor) {
			const folderPath = path.dirname(editor.document.fileName);
			const fileName = path.basename(editor.document.fileName);
			const prompt = editor.document.getText();
			const cancellationTokenSource = new vscode.CancellationTokenSource();
			await vscode.window.withProgress(
				{
					location: vscode.ProgressLocation.Notification,
					title: 'Generating unit test...',
					cancellable: true,
				},
				async (progress, token) => {
					token.onCancellationRequested(() => {
						console.log('User canceled the long running operation');
						cancellationTokenSource?.cancel();
					});
					await generateTestFile(folderPath, fileName, prompt);
					progress.report({ increment: 100 });
				}
			);
		}
	} catch (error: any) {
		if (error.response?.status === 401) {
			vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
		} else {
			vscode.window.showErrorMessage(error.message);
		}
	}
});
